File: ios/Halogen/Core/FriendlyError.swift
import Foundation

/// Human copy for every failure class — never the raw enum/NSError dump
/// (web: the classify funnel + auth error_message mapping). Every inline
/// error slot, toast, and the sync-failure surface routes through here;
/// screens with context-specific copy (ConnectView) override then delegate.
enum FriendlyError {
    static func message(_ error: Error) -> String {
        switch error {
        case let client as HalogenClient.ClientError:
            switch client {
            case .api(let fields):
                let messages = fields.values.flatMap { $0 }.compactMap(\.message)
                return messages.isEmpty
                    ? "The server rejected the request."
                    : messages.joined(separator: " · ")
            case .http(401):
                return "Session expired — sign in again."
            case .http(404):
                return "Not found on the server — it may have been removed."
            case .http(let status) where status >= 500:
                return "The server hit an error (\(status)). Try again in a moment."
            case .http(let status):
                return "The server refused the request (HTTP \(status))."
            case .offline:
                return "You're offline — reconnect to do this."
            case .signedOut:
                return "Not signed in."
            case .emptyData:
                return "The server answered with an unexpected response."
            }
        case let url as URLError:
            switch url.code {
            case .cannotFindHost, .dnsLookupFailed:
                return "Can't find the server — check the address."
            case .cannotConnectToHost, .networkConnectionLost, .timedOut:
                return "Can't reach the server right now — try again."
            case .notConnectedToInternet, .dataNotAllowed:
                return "You're offline — connect to a network first."
            case .secureConnectionFailed, .serverCertificateUntrusted,
                .serverCertificateHasBadDate, .serverCertificateHasUnknownRoot,
                .serverCertificateNotYetValid:
                return "Secure connection failed — check the server's certificate."
            case .cancelled:
                return "Cancelled."
            default:
                return "Network request failed — check the connection and try again."
            }
        case is DecodingError:
            return "Couldn't read the server's response — app and server versions may not match."
        case let embedded as HalogenCore.EmbeddedUserError:
            return embedded.description
        case let localized as LocalizedError:
            return localized.errorDescription ?? "Something went wrong — try again."
        default:
            return "Something went wrong — try again."
        }
    }
}

File: android/app/src/main/kotlin/org/fgsec/halogen/core/FriendlyError.kt
package org.fgsec.halogen.core

import java.io.IOException
import java.io.InterruptedIOException
import java.net.ConnectException
import java.net.NoRouteToHostException
import java.net.SocketException
import java.net.UnknownHostException
import javax.net.ssl.SSLException
import kotlin.coroutines.cancellation.CancellationException
import kotlinx.serialization.SerializationException
import org.fgsec.halogen.networking.HalogenClient
import uniffi.halogen_mobile.CoreException

/// Human copy for every failure class — never the raw exception dump
/// (web: the classify funnel + auth error_message mapping). Every inline
/// error slot, toast, and the sync-failure surface routes through here;
/// screens with context-specific copy (ConnectView) override then delegate.
object FriendlyError {
    fun message(error: Throwable): String = when (error) {
        is HalogenClient.ClientError -> clientMessage(error)
        // OkHttp/IO failure classes — the URLError mapping's Android shape.
        is UnknownHostException ->
            "Can't find the server — check the address."
        is SSLException ->
            "Secure connection failed — check the server's certificate."
        is ConnectException, is NoRouteToHostException, is SocketException ->
            if (error.message?.contains("unreachable", ignoreCase = true) == true)
                "You're offline — connect to a network first."
            else "Can't reach the server right now — try again."
        // SocketTimeoutException and OkHttp's callTimeout both land here.
        is InterruptedIOException ->
            "Can't reach the server right now — try again."
        is CancellationException -> "Cancelled."
        is IOException ->
            if (error.message == "Canceled") "Cancelled."
            else "Network request failed — check the connection and try again."
        is SerializationException ->
            "Couldn't read the server's response — app and server versions may not match."
        is HalogenCore.EmbeddedUserException ->
            error.message ?: "Something went wrong — try again."
        is CoreException.Server -> error.msg
        else -> "Something went wrong — try again."
    }

    private fun clientMessage(error: HalogenClient.ClientError): String = when (error) {
        is HalogenClient.ClientError.Api -> {
            val messages = error.errors.values.flatten().mapNotNull { it.message }
            if (messages.isEmpty()) "The server rejected the request."
            else messages.joinToString(" · ")
        }
        is HalogenClient.ClientError.Http -> when {
            error.code == 401 -> "Session expired — sign in again."
            error.code == 404 -> "Not found on the server — it may have been removed."
            error.code >= 500 -> "The server hit an error (${error.code}). Try again in a moment."
            else -> "The server refused the request (HTTP ${error.code})."
        }
        is HalogenClient.ClientError.Offline -> "You're offline — reconnect to do this."
        is HalogenClient.ClientError.SignedOut -> "Not signed in."
        is HalogenClient.ClientError.EmptyData ->
            "The server answered with an unexpected response."
    }
}
